### PR TITLE
[FIX JENKINS-18114] Exclude /cli URL from crumb requirement

### PR DESCRIPTION
**Pull request by daniel-beck**
[JENKINS-18114](https://issues.jenkins-ci.org/browse/JENKINS-18114)

Advice on how to test this welcome.

### CODE REVIEW

**Comment by olivergondza** [comment]
It seems to be tightly coupled with `CLIAction`, perhaps better make it an inner class.

**Comment by daniel-beck** [comment]
@olivergondza That was my first draft, but decided against it. The `CLIAction` is only one half of the protocol. Ultimately, not a big deal, as it's `@Restricted`, so can move anywhere at any time.

**Comment by olivergondza** [comment]
Agreed.

**Comment by daniel-beck** [comment]
@reviewbybees @jenkinsci/code-reviewers Would appreciate some feedback here. If it's acceptable to you I'd like to get this merged now and figure out how to test later to make CSRF protection by default less painful.

**Comment by oleg-nenashev** [comment]
Looks good to me, but actually I have no idea why would CLI need to access this page via POST. 
Maybe switching it to GET is a better solution.

Also it would be great to see a test, of course

**Comment by jglick** [comment]
I think a test is required, since from the issue description it is not at all clear to me what is going on here. Is the client trying first to use the JNLP port? If so, why is that failing—do you need to disable that to see the effect? (Something new in 2.0?)

It should be testable, at any rate, though not via `CLICommandInvoker`. Seems `CLIActionTest.testDuplexHttp` basically does it, so why is that passing?

As to why is it sending a `POST` request, see [this code](https://github.com/jenkinsci/jenkins/blob/4927968eb0c99810a414bad2d3d96bc148fd236f/cli/src/main/java/hudson/cli/FullDuplexHttpStream.java#L67-L86) for something like an explanation.

**Comment by daniel-beck** [comment]
@jglick

> Is the client trying first to use the JNLP port? If so, why is that failing—do you need to disable that to see the effect? (Something new in 2.0?)

It's an ancient issue – the CLI has an HTTP based fallback mode if connecting via JNLP port fails. Now that we're shipping Jenkins with JNLP port disabled and CSRF protection enabled by default, this becomes more relevant than it was before.

https://github.com/jenkinsci/jenkins/blob/732696bd89ac42c02b5c76d4e1cfc1796ca02bb6/cli/src/main/java/hudson/cli/CLI.java#L130

**Comment by daniel-beck** [comment]
Turns out the problem here was that the "CLI request emulation" in `CLIActionTest#serveCliActionToAnonymousUserWithoutPermissions` set the CSRF crumb for the request when the CLI doesn't actually do that.

While a test with the real CLI would be preferable, this should be close enough.

**Comment by daniel-beck** [comment]
@jglick @jenkinsci/code-reviewers Please re-review.

**Comment by daniel-beck** [comment]
@oleg-nenashev 

> I have no idea why would CLI need to access this page via POST. Maybe switching it to GET is a better solution.

[According to this comment](https://github.com/jenkinsci/jenkins/blob/38de41d14296f977901f2567f30994859a6badc3/cli/src/main/java/hudson/cli/FullDuplexHttpStream.java#L67) it's needed to circumvent browser caches. Besides, we're using `GET /cli` for the command documentation.

**Comment by stephenc** [comment]
@oleg-nenashev The CLI uses POST to support CLI usage where the TCP agent port is not accessible. Is basically does a protocol switch using the fact that POST requests have an input and output stream and sets up a remoting channel over same.

File: core/src/main/java/hudson/cli/CliCrumbExclusion.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * Makes CLI HTTP fallback work with CSRF protection enabled (JENKINS-18114).
+ */
+@Extension
+@Restricted(DoNotUse.class)
+public class CliCrumbExclusion extends CrumbExclusion {
+    @Override
+    public boolean process(HttpServletRequest request, HttpServletResponse response, FilterChain chain) throws IOException, ServletException {
+        String pathInfo = request.getPathInfo();
+        if (pathInfo != null && "/cli".equals(pathInfo)) {
```
Comment (jglick) [inline core/src/main/java/hudson/cli/CliCrumbExclusion.java]: The null check is redundant in this case.

File: core/src/main/java/hudson/cli/CliCrumbExclusion.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Makes CLI HTTP fallback work with CSRF protection enabled (JENKINS-18114).
+ */
+@Extension
+@Restricted(DoNotUse.class)
+public class CliCrumbExclusion extends CrumbExclusion {
+    @Override
+    public boolean process(HttpServletRequest request, HttpServletResponse response, FilterChain chain) throws IOException, ServletException {
+        String pathInfo = request.getPathInfo();
```
Comment (jglick) [inline core/src/main/java/hudson/cli/CliCrumbExclusion.java]: Basically every `CrumbExclusion` looks the same (though most go through a couple iterations of forgetting to distinguish `/something` vs. `/something/more` vs. `/somethingunrelated`); perhaps we should have a `@CrumbExcluded` annotation you could put on a Stapler-bound getter or `do*` method?

**Review by jglick — APPROVED** [review]
:bee:

**Pull request merged**